src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/OptionsTests.cs

```diff
@@ -661,6 +661,10 @@ public static void Options_JsonSerializerContext_Net6CompatibilitySwitch_FallsBa
                 JsonConverter converter = JsonContext.Default.Options.GetConverter(typeof(MyClass));
                 Assert.IsAssignableFrom<JsonConverter<MyClass>>(converter);
 
+                // Serialization using JsonSerializerContext now uses the reflection fallback.
+                json = JsonSerializer.Serialize(unsupportedValue, unsupportedValue.GetType(), JsonContext.Default);
```

Comment:
This is really surprising behavior.

1. Does this change cause this?  Or is this an unrelated test addition?
2. What makes this work now when it didn't work 11 lines up?

Reply:
This change is causing it. The motivation behind the PR was to remove the duplication of the SG `JsonSerializerContext.GetTypeInfo` and `IJsonTypeInfoResolver.GetTypeInfo` implementations, so this PR shims the former on top of the latter (via the `JsonSerializerOptions` cache).

A side-effect of making `JsonSerializerOptions` the source of truth for `JsonSerializerContext` metadata resolution is that it also inherits the implicit fallback behavior, provided that the .NET 6 compatibility feature switch is turned on. The implicit fallback only triggers once you call one of the `JsonSerializerOptions` overloads, which explains why you're seeing the difference in behavior in the two test lines.

I'm not particularly concerned about it, it only occurs when the compatibility switch is turned on and it actually makes the behavior consistent with how the `JsonSerializerOptions` overloads behave.